File: 0200-number-of-islands/solution.py
class UnionFind(object):
    def __init__(self, items):
        self.items = items
        self.itemMap = {items[i]:i for i in range(len(items))}
        self.rank = [0 for _ in range(len(items))]
        self.parent = [i for i in range(len(items))]
        self.groups = set(self.parent)
        
    def find(self, item):
        index = self.itemMap[item]
        if self.parent[index] != index:
            parentItem = self.items[self.parent[index]]
            self.parent[index] = self.find(parentItem)
        return self.parent[index]
    
    def union(self, item1, item2):
        
        x, y = self.find(item1), self.find(item2)
        if x == y:
            return
        if self.rank[x] < self.rank[y]:
            x,y = y,x
            
        if self.rank[x] >= self.rank[y]:
            self.parent[y] = x            
        
            if self.rank[x] == self.rank[y]:
                self.rank[x] += 1
            
            self.groups.remove(y)
               
    
    def getNumIslands(self):
        return len(self.groups)
        
        
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        
        
        steps = [(-1,0),(1,0),(0,1),(0,-1)]
        visited = set()
        ones = list()
        # islands = None

        
        def isValid(x,y):
            return 0 <= x < len(grid) and 0 <= y < len(grid[0])
        
        def dfs(x,y):
            if (x,y) not in visited:
                visited.add((x,y))
                for step in steps:
                    dx, dy = step
                    if isValid(x+dx, y + dy) and grid[x+dx][y+dy] == '1':
                        # islands.union((x,y),(x+dx,y+dy))
                        dfs(x+dx,y+dy)  
                
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    ones.append((i,j))
                    
        islands = 0

        for one in ones:
            if one not in visited:
                x,y = one
                dfs(x,y)
                islands += 1
            
        return islands
            
        
        

File: 0200-number-of-islands/solution.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        const int nr = grid.size();
        const int nc = grid[0].size();

        grid[r][c] = '0';

        if (r - 1 >= 0 && grid[r - 1][c] == '1') dfs(grid, r-1,c);
        if (r + 1 < nr && grid[r + 1][c] == '1') dfs(grid, r+1,c);
        if (c - 1 >= 0 && grid[r][c - 1] == '1') dfs(grid, r,c - 1);
        if (c + 1 < nc && grid[r][c + 1] == '1') dfs(grid, r,c + 1);

    }
    int numIslands(vector<vector<char>>& grid) {
        const int nr = grid.size();
        const int nc = grid[0].size();
        int numIslands = 0;

        for (int i = 0; i < nr; ++i) {
            for (int j = 0; j < nc; ++j) {
                if (grid[i][j] == '1') {
                    numIslands++;
                    dfs(grid, i, j);
                }
            }
        }

        return numIslands;

    }
};
